### DISCUSSION TITLE
Multitenancy

### DISCUSSION THREAD

**Discussion by mxm1070** [filamentphp/filament#1407] [Help] [answered]
Is there anyway to make this work with multitenancy? I'm using [spatie/laravel-multitenancy ](https://github.com/spatie/laravel-multitenancy) and I'm running into an issue registering resources, pages, and widgets. 

Because I'm trying keep things separate, I've split my app/Filament into app/Filament/Landlord and app/Filament/Tenant. The easiest thing I could think of was to change the paths and namespaces in the config file, but that doesn't work because I can't hard code them, and I can't make them dynamic because the tenant identification hasn't happened by the time the config file is loaded.

Next I thought I could change the config inside the AppServiceProvider boot method like so:

if(Tenant::checkCurrent())
{
    config(['filament.pages.namespace' => 'App\\Filament\\Tenant\\Pages']);
    config(['filament.pages.path' => app_path('Filament/Tenant/Pages')]);
    config(['filament.resources.namespace' => 'App\\Filament\\Tenant\\Resources']);
    config(['filament.resources.path' => app_path('Filament/Tenant/Resources')]);
    config(['filament.widgets.namespace' => 'App\\Filament\\Tenant\\Widgets']);
    config(['filament.widgets.path' => app_path('Filament/Tenant/Widgets')]);
}

but this doesn't work because the resources, pages, and widgets are discovered by Filament in the register step and not the boot step.

So then I discovered the Filament::registerResources method so in the boot step I tried:

Filament::registerResources([
    \App\Filament\Tenant\Resources\CustomerResource::class,
]);

and this throws an error: Route [filament.resources.customers.index] not defined when I try to load the dashboard.

And I'm out of ideas. It seems that all methods take place in the register step, which is too early in my multitenancy setup.

**Comment by Majkie** [comment]
Hey, just yesterday I asked Dan on how to solve same thing and we ended up with the best solution by using policies, I used stancl/tenancy (archtechx/tenancy) to test multi tenancy, so it might be a little different for you.

Since Filament resources are bound to your policies you can do this to hide and "disable" resource on tenant, but keep visible and usable on master. In case of widgets use `canView()` inside the widget where you can do the same thing.

```php
public function viewAny()
{
  if (tenancy()->initialized) {
    return false;
  }

  return true;
}
``` 

Dan also proposed to use `Filament::registerResources()` but that didn't seem to work for me, I had the same problem as you did even if I tried to register resources after tenancy was initialized properly, so I guess you'll have to stick with policies solution.

I'm currently working on article on how to use tenancy with Filament, so I'll try to look more into `Filament::registerResources()` possible solution, as it would be better than one by one checking for initialized tenant in policies. If I find some kind of possible way to deal with it, I'll post it here.

**Reply by mxm1070** [reply]
Thanks for the reply, but I don't think this is a solution. I would like to be able to keep my landlord and tenant files separate, but using a policy solution would force me to keep all the code in one file because preventing the resource/page/widget from registering in navigation doesn't not actually prevent it from being loaded. 

I can see this in ray by putting the following code in my Landlord and Tenant CustomerResource:

```php
protected static function shouldRegisterNavigation(): bool
{
    ray('landlord');
    return !config('multitenancy.tenant_model')::checkCurrent();
}

protected static function shouldRegisterNavigation(): bool
{
    ray('tenant');
    return config('multitenancy.tenant_model')::checkCurrent();
}
```
and then in ray or the log you can see:

```
17:02:27
landlord
CustomerResource.php:27

17:02:27
tenant
CustomerResource.php:26
```

And even though the Tenant CustomerResource isn't displayed in the Landlord the Tenant CustomerResource overrides the Landlord, meaning the file is still loaded and whichever loads last wins. Nova also has this problem, but in Nova you can stop the autoloading and  manually load the resources in the boot method.

As far as I can tell the only way I've been able to get this to work is to override the getModel function to return the correct Landlord or Tenant class, then split all the calls inside the Resource Class to return different schemas if each version is supposed to have a different structure. Obviously this isn't great for maintainability.

```php
if(config('multitenancy.tenant_model')::checkCurrent())
{
    return $form
        ->schema([
            //
        ]);
}
else
{
    return $form
        ->schema([
            //
        ]);
}
```

Further I run into a problem where if you have a resource that is available in the Landlord or the Tenant, but not both, you have to give it a dummy class in the getModel call or it will either through a class not found error in the context that it's not being used or fail to register the the route in the context that is it is being used. Then you have to hide it. This also falls apart if a package or something else references the static::$model instead of using the static::getModel().

```php
public static function getModel(): string
{
    return config('multitenancy.tenant_model')::checkCurrent()
        ? \App\Models\Tenant\Customer::class
        : \App\Models\Tenant\Customer::class; // '' or null doesn't work
}

protected static function shouldRegisterNavigation(): bool
{
    return config('multitenancy.tenant_model')::checkCurrent();
}
```

None of this is great, and it would all go away if we could manually load the resources/pages/widgets in the boot step instead of the register step. But this is the best idea I have because I can't figure out why Filament does it this way. I'm new to Livewire so maybe it's a Livewire problem with how routes have to be registered. Either way I would love a better solution.

**Reply by ryangjchandler** [reply]
Filament will automatically register your resources if they're inside of the app/Filament folder. Same goes for pages, etc. 

If you store them outside of that folder, you'll be able to manually register them with the various `Filament::register*` methods. 

This is definitely a niche scenario for Filament users at the moment - we haven't spent a tonne of time figuring out the nitty gritty details.

The more requests we get for this sort of thing though, the more time we'll spend thinking about documentation and alternative APIs for the use case.

**Reply by ryangjchandler** [reply]
I should also add that the reason Filament registers things in the `register` method of the service provider is so we can guarantee everything is available when we come to register routes. We've had issues in the past with moving this around, hence the rather rigid setup.

**Reply by Majkie** [reply]
I kind of made it work I believe, if you move discovery of pages, resources and widgets to bootingPackage, it won't change default behaviour, thus you can opt out of package discovery, with that you have to add FilamentServiceProvider in app.php making sure your provider where you want to register filament things has higher priority, and manually initialize tenant inside of your service provider boot method instead of middleware. Then you could just check if tenant is initialized and register related resources, widgets and pages.

```php
public function bootingPackage()
{
    $this->discoverPages();
    $this->discoverResources();
    $this->discoverWidgets();
}
```

Simple example I used that is working. (AppServiceProvider, boot)

```php
$domain = $this->makeSubdomain(request()->getHost());

if ($domain !== null) {
    /** @var Tenant $tenant */
    $tenant = config('tenancy.tenant_model')::query()
        ->whereHas('domains', function (Builder $query) use ($domain) {
            $query->where('domain', $domain);
    })
    ->with('domains')
    ->first();

    if ($tenant) {
        tenancy()->initialize($tenant);
    }
}

if (tenancy()->initialized) {
    Filament::registerResources([
        UserResource::class,
    ]);
} else {
    Filament::registerResources([
        TenantResource::class
    ]);
}
```

What do you guys think? It's not really happy solution, but at the very least it works.

**Reply by Majkie** [reply]
> I should also add that the reason Filament registers things in the `register` method of the service provider is so we can guarantee everything is available when we come to register routes. We've had issues in the past with moving this around, hence the rather rigid setup.

You don't have to worry about this thing, because `bootingPackage()` is called before `loadRoutesFrom()`, viz `PackageServiceProvider`.

**Reply by mxm1070** [reply]
Sorry, I'm a little confused about what you mean. Your idea is to publish the FilamentServiceProvider, add it to the service provider load order, and change the bootingPackage function to rediscover the various files? Do you have a working example you can link me?

**Reply by Majkie** [reply]
Not to publish FilamentServiceProvider, that wouldn't make any sense, rather to actually merge PR with that change in current FilamentServiceProvider. Then in `composer.json` add filament/filament to dont-discover and register `FilamentServiceProvider::class` in `app.php` underneath your service provider defining the logic where you initialize your tenant and register your resources in boot method, that can be done in AppServiceProvider. It's a little complicated but makes it possible to use `Filament::registerResources()` with tenancy properly. I could make a PR with this change, but it would be better if @danharrin or @ryangjchandler told me first if it's ok solution.

**Comment by mxm1070** [comment, marked as answer]
I've played around with what @Majkie [said](https://github.com/laravel-filament/filament/discussions/1407#discussioncomment-2101891) and the key seems to be to add `filament/filament` to the composer don't discover and manually register the FilamentServiceProvider at the end of the providers list. After that it works fine without auto discovery. If you want the auto discovery to work the `bootingPackage` change does work, but it's not necessary for my use. While I did not observe any problems when I tested it, it would be up to @ryangjchandler and the other maintainers to decide if that change has wider ramifications.

My process to get this to work is as follows:

1.  Add `filament/filament` to the composer.json don't discover:
```json
"extra": {
    "laravel": {
        "dont-discover": [
            "filament/filament"
        ]
    }
},
```
2. Add `Filament\FilamentServiceProvider::class` to the `config/app.php` providers array:
```php
'providers' => [
        /*
         * Application Service Providers...
         */
        App\Providers\AppServiceProvider::class,
        //...
        Filament\FilamentServiceProvider::class,
    ],
```
3. Break the auto discovery but setting the paths to null in `config/filament.php':
```php
'pages' => [
        'namespace' => 'App\\Filament\\Pages',
        'path' => null,
        'register' => [
            Pages\Dashboard::class,
        ],
    ],
'resources' => [
        'namespace' => 'App\\Filament\\Resources',
        'path' => null,
        'register' => [],
    ],
'widgets' => [
        'namespace' => 'App\\Filament\\Widgets',
        'path' => null,
        'register' => [
            Widgets\AccountWidget::class,
            Widgets\FilamentInfoWidget::class,
        ],
    ],
```
4. In `App\Providers\AppServiceProvider.php` boot manually register the components:
```php
class AppServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if(Tenant::checkCurrent())
        {
            Filament::registerResources([
                \App\Filament\Tenant\Resources\CustomerResource::class
            ]);
        }
        else
        {
            Filament::registerResources([
                \App\Filament\Landlord\Resources\CustomerResource::class
            ]);
        }
    }
}
```
5. Getting tired of that I spun my own auto discovery based on how auto discovery is done in the `Filament\FilamentServiceProvider.php`:
```php
class AppServiceProvider extends ServiceProvider
{
    public function boot()
    {
        if(Tenant::checkCurrent())
        {
            Filament::registerResources($this->registerClassesInDirectory(
                app_path('Filament/Tenant/Resources'),
                fn ($class): bool => is_subclass_of($class, Resource::class) && (!(new ReflectionClass($class))->isAbstract())
            ));

            Filament::registerPages($this->registerClassesInDirectory(
                app_path('Filament/Tenant/Pages'),
                fn ($class): bool => is_subclass_of($class, Page::class) && (!(new ReflectionClass($class))->isAbstract())
            ));

            Filament::registerWidgets($this->registerClassesInDirectory(
                app_path('Filament/Tenant/Widgets'),
                fn ($class): bool => is_subclass_of($class, Widget::class) && (!(new ReflectionClass($class))->isAbstract())
            ));
        }
        else
        {
            Filament::registerResources($this->registerClassesInDirectory(
                app_path('Filament/Landlord/Resources'),
                fn ($class): bool => is_subclass_of($class, Resource::class) && (!(new ReflectionClass($class))->isAbstract())
            ));

            Filament::registerPages($this->registerClassesInDirectory(
                app_path('Filament/Landlord/Pages'),
                fn ($class): bool => is_subclass_of($class, Page::class) && (!(new ReflectionClass($class))->isAbstract())
            ));

            Filament::registerWidgets($this->registerClassesInDirectory(
                app_path('Filament/Landlord/Widgets'),
                fn ($class): bool => is_subclass_of($class, Widget::class) && (!(new ReflectionClass($class))->isAbstract())
            ));
        }
    }

    private function registerClassesInDirectory(string $directory, ?Closure $filter = null): array
    {
        $filesystem = app(Filesystem::class);

        if($filesystem->missing($directory))
        {
            return [];
        }

        return collect($filesystem->allFiles($directory))
            ->map(function (SplFileInfo $file) use ($directory) {
                return (string) Str::of($directory)
                    ->remove(base_path().'/')
                    ->ucfirst()
                    ->append('\\', $file->getRelativePathname())
                    ->replace('/', '\\')
                    ->remove('.php');
            })
            ->when($filter, fn ($collection) => $collection->filter($filter))
            ->toArray();
    }
}
```
Everything seems to be working fine now and I can cleanly separate all of my code exactly how I want.

**Comment by bjornbjorn** [comment]
Thanks to @mxm1070 for outlining his solution, I went down that path initially but then ended up solving this in another way, basically just a change in the filament.php config.

Added this function:
```php
function tenant_aware_filament_path($path = ''): string|null
{
   if (app()->runningInConsole()) {
       return null;
   }

    if (app()->request->getHost() === env('APP_CENTRAL_DOMAIN')) {
        return app()->path('Filament/Central/'.$path);
    }

    return app()->path('Filament/Tenant/'.$path);
}
```

.. and then used this everywhere in the config like this:

```php
    'resources' => [
        'namespace' => 'App\\Filament\\Resources',
        'path' => tenant_aware_filament_path('Resources'),
        'register' => [],
    ],
```

This way Filament will load the resources from either the `Filament/Central` or `Filament/Tenant` directories. It's based on the central domain which is specified in the `.env` file.

Seems to work fine and I can login on /admin both on the central domain and the tenants and edit the different resources there.

**Reply by ryangjchandler** [reply]
@bjornbjorn Just be aware that this might affect the config file caching in production and might not work.

**Reply by bjornbjorn** [reply]
ooof, thanks for the heads up :) .. I'm currently in the "let's see what happens" stage, and that one most def would've popped up at some point. I guess either disabling caching of config files, if possible, or having the config file use some class/helper that isn't cached (but I guess the compiled config is cached so that will most likely not work). Will need to look more into that. Thanks again!
